### PR TITLE
Publish collision monitor state

### PR DESCRIPTION
**Pull request by kaichie**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #3496  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Dexory Simulation |

---

## Description of contribution in a few bullet points

* Added a new msg in `nav2_msgs` - `CollisionMonitorState.msg`
* Publish CollisionMonitor action state
* Also fixed that the first action will not be printed if it is `DO_NOTHING`
* Also updated the state change logic to check with `polygon_name` instead of `action_type` (To handle multiple `slowdown` polygons correctly)

## Description of documentation updates required from your changes

* Added new topic + param, so need to add that to default configs and documentation page

---

## Future work that may be required in bullet points

* This PR report the CM's action state change based on the input `cmd_vel`, but I think it will be very useful to publish the polygon state even when the robot is not moving too. Especially for cases like when the `stop` polygon is triggered(regardless of `cmd_vel` speed), or when there are points in `approach` polygon where the robot will not be able to move. (Probably similar requirement as mentioned [here](https://github.com/ros-planning/navigation2/pull/3500#issuecomment-1480890637))

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_collision_monitor/include/nav2_collision_monitor/types.hpp [hunk 1]
```diff
@@ -65,14 +67,16 @@ enum ActionType
   DO_NOTHING = 0,  // No action
   STOP = 1,  // Stop the robot
   SLOWDOWN = 2,  // Slowdown in percentage from current operating speed
-  APPROACH = 3  // Keep constant time interval before collision
+  APPROACH = 3,  // Keep constant time interval before collision
+  ACTION_UNKNOWN = 4 // Default for initializing action type
```
Comment (SteveMacenski) [inline nav2_collision_monitor/include/nav2_collision_monitor/types.hpp:71]: How is this semantically different from just DO_NOTHING?

File: nav2_collision_monitor/src/collision_monitor_node.cpp [hunk 2]
```diff
@@ -105,7 +109,7 @@ CollisionMonitor::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
   process_active_ = false;
 
   // Reset action type to default after worker deactivating
-  robot_action_prev_ = {DO_NOTHING, {-1.0, -1.0, -1.0}};
+  robot_action_prev_ = {ACTION_UNKNOWN, {-1.0, -1.0, -1.0}, ""};
```
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:112]: L120 below you need to deactivate the new `collision_monitor_state_pub_`

File: nav2_collision_monitor/src/collision_monitor_node.cpp [hunk 3]
```diff
@@ -480,6 +495,16 @@ void CollisionMonitor::printAction(
   }
 }
 
+void CollisionMonitor::publishActionState(const Action & robot_action) const
```
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:498]: I'd prefer if this as combined with the print function into a `notifyActionState` method logging + publishing

File: nav2_collision_monitor/include/nav2_collision_monitor/types.hpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_collision_monitor/include/nav2_collision_monitor/types.hpp:71]: If this is truly different, a default action should be `0` not `4` by convention and auto-initialization

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Looks good in general, most are nit picks and a couple of questions

File: nav2_collision_monitor/src/collision_monitor_node.cpp [hunk 4]
```diff
@@ -377,9 +386,12 @@ void CollisionMonitor::process(const Velocity & cmd_vel_in)
     }
   }
 
-  if (robot_action.action_type != robot_action_prev_.action_type) {
+  if (robot_action_prev_.action_type == ACTION_UNKNOWN ||
+    robot_action.polygon_name != robot_action_prev_.polygon_name)
```
[the remark this answers is not in the document]
Comment (kaichie) [reply nav2_collision_monitor/src/collision_monitor_node.cpp:390]: If I understand correctly, CM support multiple `slowdown` polygon with different `slowdown_ratio`, when there is a change of `slowdown` polygon from lower ratio to higher ratio, the `action_type` will stay the same but the activated polygon will be different. This is the reason I swap to `polygon_name` to reflect the correct polygon.

[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav2_collision_monitor/src/collision_monitor_node.cpp:391]: Oh ok. Lets see what @AlexeyMerzlyakov thinking but first glance that makes sense to me.

File: nav2_collision_monitor/include/nav2_collision_monitor/types.hpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_collision_monitor/include/nav2_collision_monitor/types.hpp]: I don't see why you should publish any action if the first is `DO_NOTHING` - by definition it's not doing anything. Only just to check that its running? 

I'm not against this idea, just wanting to make sure its solving a real and not a perceived problem

Comment (kaichie) [reply nav2_collision_monitor/include/nav2_collision_monitor/types.hpp]: > I don't see why you should publish any action if the first is DO_NOTHING - by definition it's not doing anything.

Ah okay, I thought I should, but if that's the case then there is no need to add this new state. I was just using it to publish the first `DO_NOTHING`.

File: nav2_collision_monitor/params/collision_monitor_params.yaml [hunk 5]
```diff
@@ -4,6 +4,7 @@ collision_monitor:
     odom_frame_id: "odom"
     cmd_vel_in_topic: "cmd_vel_raw"
     cmd_vel_out_topic: "cmd_vel"
+    collision_monitor_state_topic: "collision_monitor_state"
```
Comment (AlexeyMerzlyakov) [inline nav2_collision_monitor/params/collision_monitor_params.yaml:7]: How about to name parameter just `state_topic` (and for all variable names ditto)?

File: nav2_collision_monitor/test/collision_monitor_node_test.cpp [hunk 6]
```diff
@@ -579,6 +610,8 @@ TEST_F(Tester, testProcessStopSlowdown)
   ASSERT_NEAR(cmd_vel_out_->linear.x, 0.5 * SLOWDOWN_RATIO, EPSILON);
   ASSERT_NEAR(cmd_vel_out_->linear.y, 0.2 * SLOWDOWN_RATIO, EPSILON);
   ASSERT_NEAR(cmd_vel_out_->angular.z, 0.1 * SLOWDOWN_RATIO, EPSILON);
+  ASSERT_EQ(action_state_->action_type, SLOWDOWN);
```
Comment (AlexeyMerzlyakov) [inline nav2_collision_monitor/test/collision_monitor_node_test.cpp:613]: For correct test work, we need ensure that `action_state_` was received. For this it is need to:
1. Initialize action_state_ with `nullptr`, as made it for `cmd_vel_out_` in `Tester::publishCmdVel()` routine
2. Add new `Tester::waitActionState()` method like `Tester::waitCmdVel()`
3. Use this method before checking of `action_state_` propierties in each testcase

**Review by AlexeyMerzlyakov — COMMENTED** [review]
There are some items, I've mentioned, but in overall, this seems a good approach for me

File: nav2_collision_monitor/src/collision_monitor_node.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (AlexeyMerzlyakov) [reply nav2_collision_monitor/src/collision_monitor_node.cpp]: This change should be OK: https://github.com/ros-planning/navigation2/pull/3504#discussion_r1147726879.
We are not publishing anything more that needed, or missing something.

File: nav2_collision_monitor/src/collision_monitor_node.cpp [hunk 7]
```diff
@@ -190,6 +197,9 @@ bool CollisionMonitor::getParameters(
   nav2_util::declare_parameter_if_not_declared(
     node, "cmd_vel_out_topic", rclcpp::ParameterValue("cmd_vel"));
   cmd_vel_out_topic = get_parameter("cmd_vel_out_topic").as_string();
+  nav2_util::declare_parameter_if_not_declared(
+    node, "collision_monitor_state_topic", rclcpp::ParameterValue("collision_monitor_state"));
+  collision_monitor_state_topic = get_parameter("collision_monitor_state_topic").as_string();
```
[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav2_collision_monitor/src/collision_monitor_node.cpp]: I disagree, we should always have a logical default for topics.

File: nav2_collision_monitor/src/collision_monitor_node.cpp [hunk 8]
```diff
@@ -77,6 +82,9 @@ CollisionMonitor::on_activate(const rclcpp_lifecycle::State & /*state*/)
 
   // Activating lifecycle publisher
   cmd_vel_out_pub_->on_activate();
+  if (state_pub_ != nullptr) {
```
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:85]:
```suggestion
  if (state_pub_) {
```

Is enough - its the same but cleaner. Same for deactivate.

File: nav2_collision_monitor/src/collision_monitor_node.cpp [hunk 9]
```diff
@@ -478,6 +497,15 @@ void CollisionMonitor::printAction(
       get_logger(),
       "Robot to continue normal operation");
   }
+
+  if (state_pub_ != nullptr) {
```
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:501]: Ditto

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Otherwise LGTM, just need the complimentary documentation updates for the new parameter in the config guide + migration guide for the new feature

**Comment by kaichie** [comment]
Thanks the review @SteveMacenski and @AlexeyMerzlyakov! Created a [PR](https://github.com/ros-planning/navigation.ros.org/pull/405) for updating the config guide + migration guide too.

**Review by SteveMacenski — APPROVED** [review]
LGTM. On Alexey's final call to merge.

**Review by AlexeyMerzlyakov — APPROVED** [review]
Change reviewed: OK.
Local `colcon test --packages-select nav2_collision_monitor`: OK
Codecov test: OK
Local TB3 simulation verification: OK (CM is working well and publishing a new topic)

**Pull request merged**